Implement Multi Select Values

hahmadia opened the pull request:
#### Summary
<!--
A description of what this pull request does.
-->
Implement multi select values

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
Fixes GH-110


![image](https://user-images.githubusercontent.com/17804942/118546701-4fc6c900-b726-11eb-81d8-0b235918cb8a.png)

![image](https://user-images.githubusercontent.com/17804942/118546748-62410280-b726-11eb-9cdb-9db75cca0e46.png)

![image](https://user-images.githubusercontent.com/17804942/118548921-d7153c00-b728-11eb-9c7a-66edbebb2a0f.png)

![image](https://user-images.githubusercontent.com/17804942/118548970-e1cfd100-b728-11eb-9376-c9637ff92929.png)


![image](https://user-images.githubusercontent.com/17804942/118550482-bcdc5d80-b72a-11eb-84ee-42a8df663c73.png)




### Requirements:
* [x] Able to add a multi-select property type
* [x] Able to set multiple values on a multi-select property type
* [ ] Able to filter by multi-select property types
* [x] Multi-select property types function in all views

chenilim commented:
Minor thing: add a margin property to the multi-select values in the Card Summary and read-only card detail view. Right now the values appear joined together:
<img width="278" alt="multiselect properties" src="https://user-images.githubusercontent.com/46905241/118567145-1ee89300-b72a-11eb-8a9d-65ee0f28ab47.png">

This can be done in a subsequent commit.

chenilim commented on webapp/src/blocks/card.ts:
```diff
@@ -7,7 +7,7 @@ import {IBlock, MutableBlock} from './block'
     readonly icon: string
     readonly isTemplate: boolean
-    readonly properties: Readonly<Record<string, string>>
+    readonly properties: Readonly<Record<string, string | string[]>>
```
One thing to be careful about is supporting changing the type of a property from Select <> Mult-select. Going from single-select to multi-select should appear seamless to the user (though there may be quite involved conversion code).

Going from multi-select to single-select should retain the options, but make the value the first value in the multi-select list.

harshilsharma63 replied:
@chenilim on the same topic, I think we should also show a confirmation model when we convert property types between incompatible types. For example, a `person` type to an `updated at` type. This will avoid any surprises and data loss.

harshilsharma63 replied:
When switching between `Single Select` and `Multi Select`, should we be retain the options? Right now we're not.

chenilim replied:
Yes, to clarify:
* When changing from single-select to multi-select: The options and selected value should remain the same (only change is the user can specify additional values).
* When changing from multi-select to single-select: Retain the options, but make the value the first value in the multi-select list.

I'll leave it to @hahmadia, but both of these can be done in a subsequent checkin if necessary.

hahmadia replied:
Done

harshilsharma63 commented:
It doesn't look very clear to me if an `x` removes the chip on the left to it or the right to it. 

<img width="969" alt="Screenshot 2021-05-18 at 3 42 22 PM" src="https://user-images.githubusercontent.com/18575143/118633962-a77d3700-b7ef-11eb-9b59-ebad15cb420b.png">

We already have this component in Compass Components actually and we can tweak the CSS to bring that `x` inside the chip - https://zeroheight.com/29be2c109/p/02dffc-chips

This need a design decision though, cc @michaelgamble

harshilsharma63 commented on webapp/src/components/propertyValueElement.tsx:
```diff
@@ -56,6 +56,58 @@ const PropertyValueElement = (props:Props): JSX.Element => {
         }
     }
 
+    if (propertyTemplate.type === 'multiSelect') {
```
It would be cleaner to have an individual React component for each property type, like I did here - https://github.com/mattermost/focalboard/pull/393

It's a team call though.

chenilim replied:
You guys can make the call, but my advice is to refactor later. i.e. either check in new functionality, or refactor existing functionality. If the change is relatively small or contained, then go ahead and refactor now, otherwise my advice is to do it in a subsequent commit.

hahmadia replied:
Done

hahmadia commented:
Added margins so it doesn't look stuck
![image](https://user-images.githubusercontent.com/17804942/118835579-224a5e80-b891-11eb-9718-14bd92b24224.png)

sbishel commented:
A couple things when comparing to Notion. As suggested above. They have the remove "x" within the coloring. 

Also, they display all the options even the ones selected. This allows you to change colors or edit one that is selected. 
![Screen Shot 2021-05-21 at 3 48 30 PM](https://user-images.githubusercontent.com/12704875/119201814-d80ddc80-ba4c-11eb-8657-954701b4150c.png)

<img width="803" alt="Screen Shot 2021-05-21 at 3 54 09 PM" src="https://user-images.githubusercontent.com/12704875/119201782-c9272a00-ba4c-11eb-8cb4-cd4c8e17bdb3.png">

sbishel reviewed (changes requested):
A couple things when comparing to Notion. As suggested above. They have the remove "x" within the coloring. 

Also, they display all the options even the ones selected. This allows you to change colors or edit one that is selected. They also leave the drop down open rather than closing after each selection. 

<img width="803" alt="Screen Shot 2021-05-21 at 3 54 09 PM" src="https://user-images.githubusercontent.com/12704875/119201782-c9272a00-ba4c-11eb-8cb4-cd4c8e17bdb3.png">

![Screen Shot 2021-05-21 at 3 48 30 PM](https://user-images.githubusercontent.com/12704875/119201814-d80ddc80-ba4c-11eb-8657-954701b4150c.png)

Also the text and button should be center and even with each other.
![Screen Shot 2021-05-21 at 3 45 57 PM](https://user-images.githubusercontent.com/12704875/119202124-684c2180-ba4d-11eb-9cbd-19e373b5c910.png)

hahmadia commented:
To-Do:
- X within label
- Keep menu open after choosing an option
- Text and three dots be aligned in dropdown
- Add test for PropertyValueElement

harshilsharma63 commented:
Some issues I found -

1. The dropdown container doesn't expand all the way to contain the option list if the options are more than a few values. This happens for both single and multi select so can be fixed separate from this ticket.
<img width="791" alt="Screenshot 2021-05-26 at 12 13 26 PM" src="https://user-images.githubusercontent.com/18575143/119614328-c7c67a80-be1b-11eb-9bb4-2aac03b4fc33.png">

1. There is no way to delete a really long value. It overflows from the card, making the `x` disappear -
<img width="778" alt="Screenshot 2021-05-26 at 12 15 29 PM" src="https://user-images.githubusercontent.com/18575143/119614580-1116ca00-be1c-11eb-8e39-cfc78f99ccd2.png">

One solution is to truncate the text to a max-length and show the full text as a tooltip on hover-
<img width="691" alt="Screenshot 2021-05-26 at 12 20 34 PM" src="https://user-images.githubusercontent.com/18575143/119615209-c6e21880-be1c-11eb-9958-312e452563ec.png">

This is also an existing issue and can be seen in single select, we can fix it separately.

hahmadia commented on webapp/src/viewModel/boardTree.ts:
```diff
@@ -216,7 +216,9 @@ class MutableBoardTree implements BoardTree {
                         }
-                    } else if (propertyValue.toLowerCase().includes(searchText)) {
+
+                    // TODO: Add search capability for multi-select values BIG BOYY
```
I was planning to do it as a separate PR excuse my PR comment lol. I was hoping to get the initial implementation in and worry about the searching separate as a way to not to make PR too big.

hahmadia commented on webapp/src/components/propertyValueElement.tsx:
```diff
@@ -56,6 +59,53 @@ const PropertyValueElement = (props:Props): JSX.Element => {
     }
 
+    if (propertyTemplate.type === 'multiSelect') {
+        const values = Array.isArray(propertyValue) ?
```
Good call. I've made more abstractions.

sbishel commented on webapp/src/mutator.ts:
```diff
@@ -392,10 +396,33 @@ class Mutator {
             // Map select to their values
             for (const card of boardTree.allCards) {
-                const oldValue = card.properties[propertyTemplate.id]
+                const oldValue = Array.isArray(card.properties[propertyTemplate.id]) ? card.properties[propertyTemplate.id][0] : card.properties[propertyTemplate.id]
```
Do you need to check length here, as done above?

hahmadia replied:
Yep! I don't think it hurts to have it. I believe I missed it here. Good catch.

hahmadia replied:
This changed helped me realize and removed the entire if block above!

sbishel reviewed (changes requested):
I noticed this issue in table view. 

![Screen Shot 2021-05-26 at 4 22 41 PM](https://user-images.githubusercontent.com/12704875/119739694-d6268d80-be3f-11eb-9b83-bfec8c812db1.png)

hahmadia commented:
@harshilsharma63 

Thanks for the help. I've addressed the long text issue with your help. I also applied our fix to card view because i noticed it overflows in this view as well. Thanks for that.
![image](https://user-images.githubusercontent.com/17804942/119858941-eba0c380-bee2-11eb-81c9-15c771d5fc63.png)


@sbishel 
I've addressed all the concerns except the one where the row is slightly larger due to other ones. It seems to be caused due to padding in between the options and as well as the icon. Do you feel it must be fixed? I personally think it's fine having a slightly bigger row when you have multi select option and don't feel it takes away from anything. 

No longer see the white spacing stuff as shown in your photo above
![image](https://user-images.githubusercontent.com/17804942/119859205-2c98d800-bee3-11eb-8e6e-ad5ce6581ac5.png)


Thanks!

sbishel commented:
> @sbishel
> I've addressed all the concerns except the one where the row is slightly larger due to other ones. It seems to be caused due to padding in between the options and as well as the icon. Do you feel it must be fixed? I personally think it's fine having a slightly bigger row when you have multi select option and don't feel it takes away from anything.

Yes, I think this should be fixed. I find it distracting and sloppy. My guess is you will need to different "padding" or "margin" settings when displayed in table view to keep it the same size as a single select column.

hahmadia commented:
> > @sbishel
> > I've addressed all the concerns except the one where the row is slightly larger due to other ones. It seems to be caused due to padding in between the options and as well as the icon. Do you feel it must be fixed? I personally think it's fine having a slightly bigger row when you have multi select option and don't feel it takes away from anything.
> 
> Yes, I think this should be fixed. I find it distracting and sloppy. My guess is you will need to different "padding" or "margin" settings when displayed in table view to keep it the same size as a single select column.

Hi Scott,

Feel free to check now. The table height changing with multiselect should be fixed now. Let me know what you think.

harshilsharma63 reviewed (approved):
LGTM 👍

hahmadia commented:
I've created a card for testing to be done in the future.

https://focalboard-community.octo.mattermost.com/workspace/qgsck6cts3fwpqwyjiupjm5cde?id=47aa9bb4-6967-4a96-83c7-11bd6b20f1eb&v=baa742b3-f3f1-406e-811a-00f18d723781&c=d43904fb-8d02-4a5b-8892-e9176a1d1193

sbishel commented:
You may want to update the screenshots, since it has changed.

chenilim reviewed (approved):
I did a quick smoke test on a private build, and it looks great! 👍

hahmadia commented:
@hahmadia 

a note to self:

Update original post to include updated screenshots of work!
